---
title: "Compute per-homolog escape"
author: "Tyler Starr"
date: "12/07/2020"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---

This notebook compares single-mutant escape as measured in the SARS-CoV-2 background to the homologs that do or do not escape binding by mAbs.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra","bio3d")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
invisible(lapply(packages, library, character.only=T))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$compare_mut_dir)){
  dir.create(file.path(config$compare_mut_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Data input

Read in table of escape scores for mutants as measured in the SARS-CoV-2 background and wildtype sarbecovirus homologs, and an alignment of sarbecovirus homologs. Strip alignment columns that are gap in SARS-CoV-2

```{r input_data}
dt <- data.table(read.csv(file=config$escape_fracs_homologs,stringsAsFactors = F))
dt_mut <- data.table(read.csv(file=config$escape_muts,stringsAsFactors = F))
alignment <- read.fasta(file=config$sarbecovirus_alignment)
alignment$ali <- alignment$ali[,alignment$ali[grep("SARS-CoV-2",alignment$id),]!="-"]
```

## Build table of escape for each homolog variant

Build a long-format table listing, for each homolog amino acid state that differs from SARS-CoV-2, the escape fraction from each antibody for that variant as measured in the SARS-CoV-2 background.

```{r build_table}
variants <- data.table(homolog=factor(x=character(),levels=c(config$EurAf_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)),antibody=character(),position=numeric(),wildtype=character(),mutation=character(),mut_escape=numeric())
for(target in levels(variants$homolog)){
  if(length(grep(target,alignment$id))==1){
    alignment.row <- grep(target,alignment$id)
    for(j in 1:ncol(alignment$ali)){
      if(alignment$ali[alignment.row,j] != alignment$ali[grep("SARS-CoV-2",alignment$id),j]){
        pos <- j + 330
        wt <- alignment$ali[grep("SARS-CoV-2",alignment$id),j]
        mut <- alignment$ali[alignment.row,j]
        for(mAb in config$antibodies_ordered){
          escape <- dt_mut[condition==mAb & site==pos & mutation==mut,mut_escape]
          if(length(escape)==1){
            variants <- rbind(variants, list(target,mAb,pos,wt,mut,escape))
          }else{
            variants <- rbind(variants, list(target,mAb,pos,wt,mut,as.numeric(NA)))
          }
          
        }
      }
    }
  }else{
    print(paste(target,"pattern not in alignment exactly once"))
  }
}

```
Jitter plot to show escape frac of mutations that differ for each homolog.

```{r plot_homolog_mut_escapes, echo=T, fig.width=12, fig.height=18, fig.align="center", dpi=300,dev="png"}
# #add value indicating the heuristic cutoff for "strong" escape for each mAb
# cutoff_LYCoV016 <- max(5*median(temp[antibody=="LY-CoV016",site_total_escape]),0.05*max(temp[antibody=="LY-CoV016",site_total_escape]))
# cutoff_REGN10933 <- max(5*median(temp[antibody=="REGN10933",site_total_escape]),0.05*max(temp[antibody=="REGN10933",site_total_escape]))
# cutoff_REGN10987 <- max(5*median(temp[antibody=="REGN10987",site_total_escape]),0.05*max(temp[antibody=="REGN10987",site_total_escape]))
# 
# temp[antibody=="LY-CoV016",cutoff:=cutoff_LYCoV016]
# temp[antibody=="REGN10933",cutoff:=cutoff_REGN10933]
# temp[antibody=="REGN10987",cutoff:=cutoff_REGN10987]

#set factor for plotting order of mAbs
variants$antibody <- factor(variants$antibody,levels=config$antibodies_ordered)

#sitewise
ggplot(variants,aes(x=homolog,y=mut_escape))+
  geom_jitter(width=0.15, alpha=0.5,height=0,size=2.25,shape=16)+
  #geom_hline(data=unique(temp[,.(antibody,cutoff)]),aes(yintercept=cutoff),linetype="dashed",color="gray50")+
  theme_classic()+
  theme(axis.text.x=element_text(angle=90,hjust=1))+
  xlab("")+ylab("mut escape")+
  facet_wrap(~antibody,ncol=1)

invisible(dev.print(pdf, paste(config$compare_mut_dir,"/homolog_muts_escape.pdf",sep="")))
```

