---
title: "Compute per-homolog escape"
author: "Tyler Starr"
date: "11/20/2020"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---

This notebook filters the per-barcode escape fraction estimates, and computes summary escape fractions for each homolog.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
invisible(lapply(packages, library, character.only=T))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$escape_scores_dir)){
  dir.create(file.path(config$escape_scores_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Data input

Read in table of escape scores for barcodes, also annotated with variant counts and expression that we'll use in filtering. Remove barcodes with pre_count of 0 or NA expression effects -- these are likely barcodes that either do not express well, are generally low-frequency, or were actually "mutant" variants (but no mutation in the SSM) which were not part of the sub-pool of the homologs experiments that we used for these antibody selections.

```{r input_data}
dt <- data.table(read.csv(file=config$escape_fracs_barcodes,stringsAsFactors = F))
dt <- dt[pre_count>0 & !is.na(expression),]
```

## Filtering

We will use the per-barcode pre_count and expression scores to filter out escape scores used in computing per-homolog escape. First, let's look at the distribution of pre-counts across barcodes. Vertical lines indicate the 10th percentile of pre-count where we will filter.

```{r hist_pre_count, fig.width=4, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(log10(dt$pre_count),col="gray50",main="",xlab="pre-selection sequencing count (log10 scale)");abline(v=log10(quantile(dt$pre_count,0.1)),col="red",lty=2)

```
Let's also see how escape fraction correlates with pre-count. Theoretically, escape_fraction should not exceed 1, though we see it does, particularly when pre_count is lower.

```{r scatter_pre_count, fig.width=4, fig.height=4, fig.align="center", dpi=300,dev="png"}
plot(dt$pre_count,dt$escape_frac,pch=16,col="#00000060",log="x",xlab="pre-count (log scale)",ylab="escape fraction (shouldn't exceed 1)")
abline(h=1,lty=2,col="red")
abline(v=quantile(dt$pre_count,0.1),lty=2,col="red")
```

Remove barcode measurements for those with pre-counts in the lowest ten percentile. This corresponds to removing variants with less than `r quantile(dt$pre_count,0.1)` pre-sort counts.

```{r filter_precount}
dt <- dt[pre_count > quantile(dt$pre_count,0.1),]
```

Censor NA measurements for AncSARS1a_alt and HKU3-8 which were generally non-expressing (and so any barcodes that eke through are probably poorly expressed/artefactual)

```{r filter_nonexpressing_homologs}
dt[target%in%c("AncSARS1a_alt","HKU3-8"),escape_frac:=NA]
```

## Escape fraction per homolog

Next, let's visualize the escape fraction scores per barcode grouped by variant, with violin plots.

```{r escape_frac_vioplot, fig.width=13, fig.height=20, fig.align="center", dpi=300,dev="png"}
#set factor order for homologs to display
dt$target <- factor(dt$target, levels=config$targets_ordered)

ggplot(dt,aes(x=target,y=escape_frac))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("escape frac by homolog")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))+
  facet_wrap(~antibody,ncol=1)

```

Collapse each homolog escape fraction to its median across barcodes.

```{r collapse_median}
dt[,median_escape_frac:=median(escape_frac,na.rm=T),by=c("library","target","antibody")]
dt[,n_barcodes:=sum(!is.na(escape_frac)),by=c("library","target","antibody")]
dt[,median_expression:=median(expression,na.rm=T),by=c("library","target")]
dt_collapse <- unique(dt[,.(library,target,median_expression,antibody,median_escape_frac,n_barcodes)])
dt_collapse[,expression:=median_expression];dt_collapse[,escape_frac:=median_escape_frac]
dt_collapse <- dt_collapse[,.(library,target,expression,antibody,escape_frac,n_barcodes)]
dt_collapse[is.na(escape_frac),n_barcodes:=NA]
```

Histogram below shows that a few medians end up above the theoretical max escape fraction of 1. We set these to the maximum 1.

```{r hist_median_unaltered, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(dt_collapse$escape_frac,main="",col="gray50",xlab="homolog escape fraction",breaks=20)

dt_collapse[escape_frac>1,escape_frac:=1]
```

Also make histograms showing the typical number of barcodes on which a homolog escape fraction was averaged across. The median number of barcodes across all homolog escape fracs is `r median(dt$n_barcodes,na.rm=T)`.

```{r hist_n_barcode, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(dt_collapse$n_barcodes,main="",col="gray50",xlab="number of barcodes",breaks=20)
```

## Heatmaps

Last, make heatmaps illustrating the fraction escape of each homolog versus each antibody. These will be what we'll eventually align to the phylogeny as our data display?

First, for all homologs in the library, including ancestors. This is the default ordering of homologs set by the factor variable.

```{r heatmap_homolog-escape_all, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
#set antibody order as factor from config
dt_collapse$antibody <- factor(dt_collapse$antibody,levels=config$antibodies_ordered)

ggplot(dt_collapse,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_all.pdf",sep="")))
```

Extant homologs.

```{r heatmap_homolog-escape_extant, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
extant <- c(config$EurAf_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp <- dt_collapse[target %in% extant,];temp$target <- factor(temp$target,levels=extant)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_extant.pdf",sep="")))
```

MAP ancestors.

```{r heatmap_homolog-escape_MAP-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_MAP-ancestors.pdf",sep="")))
```

MAP plus alternative ancestors.

```{r heatmap_homolog-escape_all-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP_v_alt)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_all-ancestors.pdf",sep="")))
```

## Output

Save our final per-homolog escape fraction estimates for each antibody.

```{r output_data}
dt_collapse %>%
  mutate_if(is.numeric, round, digits=5) %>%
  write.csv(file=config$escape_fracs_homologs, row.names=F)

```

