---
title: "Compute per-homolog escape"
author: "Tyler Starr"
date: "11/20/2020"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---

This notebook filters the per-barcode escape fraction estimates, and computes summary escape fractions for each homolog.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
invisible(lapply(packages, library, character.only=T))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$escape_scores_dir)){
  dir.create(file.path(config$escape_scores_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Data input

Read in table of escape scores for barcodes, also annotated with variant counts and expression that we'll use in filtering. Remove barcodes with pre_count of 0 or NA expression effects -- these are likely barcodes that either do not express well, are generally low-frequency, or were actually "mutant" variants (but no mutation in the SSM) which were not part of the sub-pool of the homologs experiments that we used for these antibody selections.

```{r input_data}
dt <- data.table(read.csv(file=config$escape_fracs_barcodes,stringsAsFactors = F))
dt <- dt[pre_count>0 & !is.na(expression),]
```

## Filtering

We will use the per-barcode pre_count and expression scores to filter out escape scores used in computing per-homolog escape.

First, let's look at the distribution of pre-counts across barcodes. The median pre-count is `r median(dt$pre_count)`. Vertical lines on the two plots below indicate a threshold for pre_count of 1/2 that of the median pre-count, which is what we'll apply below.

```{r hist_pre_count, fig.width=4, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(log10(dt$pre_count),col="gray50",main="",xlab="pre-selection sequencing count (log10 scale)");abline(v=log10(0.5*median(dt$pre_count)),col="red",lty=2)

```
Let's also see how escape fraction correlates with pre-count. Theoretically, escape_fraction should not exceed 1, though we see it does, particularly when pre_count is lower.

```{r scatter_pre_count, fig.width=4, fig.height=4, fig.align="center", dpi=300,dev="png"}
plot(dt$pre_count,dt$escape_frac,pch=16,col="#00000060",log="x",xlab="pre-count (log scale)",ylab="escape fraction (shouldn't exceed 1)")
abline(h=1,lty=2,col="red")
abline(v=0.5*median(dt$pre_count),lty=2,col="red")
```

Remove barcode measurements for those where pre-count is less than half the median pre-count. This corresponds to removing variants with less than `r 0.5*median(dt$pre_count)` pre-sort counts.

```{r filter_precount}
dt <- dt[pre_count > 0.5*median(dt$pre_count),]
```

Censor NA measurements for AncSARS1a_alt and HKU3-8 which were generally non-expressing from our prior measurements (and so any barcodes that eke through are probably poorly expressed/artefactual)

```{r filter_nonexpressing_homologs}
dt[target%in%c("AncSARS1a_alt","HKU3-8"),escape_frac:=NA]
dt[target%in%c("AncSARS1a_alt","HKU3-8"),expression:=NA]
```

## Escape fraction per homolog

Next, let's visualize the escape fraction scores per barcode grouped by variant, with violin plots.

```{r escape_frac_vioplot, fig.width=13, fig.height=20, fig.align="center", dpi=300,dev="png"}
#set factor order for homologs to display
dt$target <- factor(dt$target, levels=config$targets_ordered)

ggplot(dt,aes(x=target,y=escape_frac))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("escape frac by homolog")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))+
  facet_wrap(~antibody,ncol=1)

```

Collapse each homolog escape fraction to its median across barcodes.

```{r collapse_median}
dt[,median_escape_frac:=median(escape_frac,na.rm=T),by=c("library","target","antibody")] #median escape across all barcodes for a homolog
dt[,n_barcodes:=sum(!is.na(escape_frac)),by=c("library","target","antibody")] #the number of barcodes on which a homolog median escape was calculated
dt[,median_expression:=median(expression,na.rm=T),by=c("library","target")] #the average expression of that homolog
dt_collapse <- unique(dt[,.(library,target,median_expression,antibody,median_escape_frac,n_barcodes)]) #collapse down to homolog-level data instead of barcode-level
dt_collapse[,expression:=median_expression];dt_collapse[,escape_frac:=median_escape_frac] #rename median quantities to just now be the homolog-level quantity
dt_collapse <- dt_collapse[,.(library,target,expression,antibody,escape_frac,n_barcodes)]
dt_collapse[is.na(escape_frac),n_barcodes:=NA]
```

Histogram below shows that a few medians end up above the theoretical max escape fraction of 1. We set these to the maximum 1.

```{r hist_median_unaltered, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(dt_collapse$escape_frac,main="",col="gray50",xlab="homolog escape fraction",breaks=20)

dt_collapse[escape_frac>1,escape_frac:=1]
```

Also make histograms showing the typical number of barcodes on which a homolog escape fraction was averaged across. The median number of barcodes across all homolog escape fracs is `r median(dt$n_barcodes,na.rm=T)`.

```{r hist_n_barcode, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
hist(dt_collapse$n_barcodes,main="",col="gray50",xlab="number of barcodes",breaks=20)
```

How does escape frac for a homolog tend to correlate with its expression? Not that this was a particularly well-devised check, but for most homologs (especially expression around 10), doesn't seem to be an overall association between expression and escape. Perhaps notable that the only measurement that really hits the "intermediate" range of escape frac (instead of being more binary 0/1) is the more poorly expressing variant AncClade2_alt1_subs_only (that is, a variant that we do expect to be perhaps less stable because of epistatic incompatibilites -- this variant contains all of the 48 substitutions that occur between AncAsia and AncClade2, but does not contain the two deletions that also occur along this branch).

If this variant were being disproportionately selected *out* by the RBD+ gating during the sort (because it is lower expressing than the average homolog), this would have the effect of dragging down its theoretical maximum escape fraction, because its theoretical maximum antibody-escape bin counts cannot reach the relative frequency of its pre-sort counts, because some fraction of the cells containing those barcodes aren't even making it through the RBD+ gate and then into the antibody-escape bin.

Consistent with this premise, it is even more notable that the four antibody escape samples for this genotype that *do* have these intermediate escape fractions (~0.5) are the four antibodies that were run together on the Aria 2-2 on the 11/9/2020 sort day -- the five antibodies run on 11/6, and the one antibody (S2X227) that were run on the Aria 2-3 on 11/9 (so even prepped in parallel, but different sorter with differently drawn sort gates). The FACS log from the 11/9 2-2 does indeed show that the RBD+ gating was drawn more stringently than in the other two batches. So, taken together, I do believe quite strongly that these intermediate escape fractions are not distinguishable from escape fractions of 1. And therefore, I wonder across all of the homologs whether we should even display binding as a quantitative 0 to 1 scale, versus just binary yes/no? Even though from the FACS log, there *were* instances where we could see clouds of cells representing different homologs with differing levels of escape (which we were hoping we could then maybe see in the phenotypes some small and meaningful differences from 0 or 1), I do not think we can effectively distinguish those genotypes when conflated by variance in expression and inherent variance in stringency of RBD+ gating across sort batches. Even though in this scatterplot, we can for the maximally expressing homologs see some variation in the escape scores away from ==0 or ==1 that might be interpretable in the end, I don't see a super principled way of how to de-conflate varying expression effects here.

```{r scatter_escape_v_expression, fig.width=4, fig.height=4, fig.align="center", dpi=300,dev="png"}
plot(dt_collapse$expression,dt_collapse$escape_frac,pch=16,col="#00000066",xlab="homolog expression",ylab="homolog escape fraction")
```

Let's also collapse to a binary 0/1 escape score indicator, in light of the observation above. I think if we only use the actual sampled homologs and don't use the inferred ancestors (which includes this expression=8 variant), we probably can keep the 0 to 1 continuous scale as these all have good expression, although I'm not sure in the end if there is valuable information in the fine-grained differences in binding score or not.

```{r binary_escape}
dt_collapse[escape_frac<0.2,binary_escape_frac:=0]
dt_collapse[escape_frac>=0.2,binary_escape_frac:=1]
```

## Heatmaps

Last, make heatmaps illustrating the fraction escape of each homolog versus each antibody. These will be what we'll eventually align to the phylogeny as our data display?

First, for all homologs in the library, including ancestors. This is the default ordering of homologs set by the factor variable.

```{r heatmap_homolog-escape_all, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
#set antibody order as factor from config
dt_collapse$antibody <- factor(dt_collapse$antibody,levels=rev(config$antibodies_ordered))

ggplot(dt_collapse,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_all.pdf",sep="")))
```

Extant homologs.

```{r heatmap_homolog-escape_extant, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
extant <- c(config$EurAf_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp <- dt_collapse[target %in% extant,];temp$target <- factor(temp$target,levels=extant)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_extant.pdf",sep="")))
```

Invert colors (so blacker indicates 'binding')

```{r heatmap_homolog-escape_extant_inverse_color, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
extant <- c(config$EurAf_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp <- dt_collapse[target %in% extant,];temp$target <- factor(temp$target,levels=extant)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(high="white",low="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_extant_inverse_colors.pdf",sep="")))
```


MAP ancestors.

```{r heatmap_homolog-escape_MAP-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_MAP-ancestors.pdf",sep="")))
```

MAP plus alternative ancestors.

```{r heatmap_homolog-escape_all-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP_v_alt)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/heatmap_homologs_all-ancestors.pdf",sep="")))
```

Same heatmaps, but with the binary 0/1 escape scores

```{r binary_heatmap_homolog-escape_all, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
#set antibody order as factor from config
ggplot(dt_collapse,aes(target,antibody))+geom_tile(aes(fill=binary_escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/binary_heatmap_homologs_all.pdf",sep="")))
```

Extant homologs.

```{r binary_heatmap_homolog-escape_extant, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
extant <- c(config$EurAf_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp <- dt_collapse[target %in% extant,];temp$target <- factor(temp$target,levels=extant)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=binary_escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/binary_heatmap_homologs_extant.pdf",sep="")))
```

MAP ancestors.

```{r binary_heatmap_homolog-escape_MAP-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=binary_escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/binary_heatmap_homologs_MAP-ancestors.pdf",sep="")))
```

MAP plus alternative ancestors.

```{r binary_heatmap_homolog-escape_all-anc, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=T}
ancestors <- c(config$ancestors_MAP_v_alt)

temp <- dt_collapse[target %in% ancestors,];temp$target <- factor(temp$target,levels=ancestors)

ggplot(temp,aes(target,antibody))+geom_tile(aes(fill=binary_escape_frac),color="black",lwd=0.1)+
  scale_fill_gradient(low="white",high="#353D41",limits=c(0,1),na.value="cyan")+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,face="bold"))

invisible(dev.print(pdf, paste(config$escape_scores_dir,"/binary_heatmap_homologs_all-ancestors.pdf",sep="")))
```

## Output

Save our final per-homolog escape fraction estimates for each antibody.

```{r output_data}
dt_collapse %>%
  mutate_if(is.numeric, round, digits=5) %>%
  write.csv(file=config$escape_fracs_homologs, row.names=F)

```

